ydb/core/tx/datashard/datashard_ut_write.cpp

```diff
@@ -716,5 +719,452 @@ Y_UNIT_TEST_SUITE(DataShardWrite) {
+                /* sending */ shards,
+                /* receiving */ shards,
+                [&](const TWriteRequestPtr& req) {
+                    // We use a lock that should have never existed to simulate a broken lock
+                    if (shardsWithBrokenLocks.contains(i)) {
+                        auto* kqpLock = req->Record.MutableLocks()->AddLocks();
+                        kqpLock->SetLockId(txId);
+                        kqpLock->SetDataShard(shardId);
+                        kqpLock->SetGeneration(1);
+                        kqpLock->SetCounter(1);
+                        kqpLock->SetSchemeShard(tableId.PathId.OwnerId);
+                        kqpLock->SetPathId(tableId.PathId.LocalPathId);
+                    }
+                });
+
+            ++expectedResults;
+        }
+
+        size_t observedReadSets = 0;
+        auto observeReadSets = runtime.AddObserver<TEvTxProcessing::TEvReadSet>(
+            [&](TEvTxProcessing::TEvReadSet::TPtr&) {
+                ++observedReadSets;
+            });
+
+        SendProposeToCoordinator(runtime, sender, txId, coordinator, minStep, maxStep, shards);
+
+        Cerr << "========= Waiting for write results =========" << Endl;
+        for (size_t i = 0; i < expectedResults; ++i) {
+            auto ev = runtime.GrabEdgeEventRethrow<NEvents::TDataEvents::TEvWriteResult>(sender);
+            auto* msg = ev->Get();
+            auto it = expectedShardStatus.find(msg->Record.GetOrigin());
+            if (it != expectedShardStatus.end()) {
+                UNIT_ASSERT_VALUES_EQUAL(msg->Record.GetStatus(), it->second);
+            } else {
+                UNIT_ASSERT_VALUES_EQUAL(msg->Record.GetStatus(), expectedStatus);
+            }
+        }
+
+        UNIT_ASSERT_VALUES_EQUAL(observedReadSets, expectedReadSets);
+    }
+
+    Y_UNIT_TEST(UpsertNoLocksArbiter) {
+        TPortManager pm;
+        TServerSettings serverSettings(pm.GetPort(2134));
+        serverSettings.SetDomainName("Root")
+            .SetUseRealThreads(false);
+
+        auto [runtime, server, sender] = TestCreateServer(serverSettings);
+
+        UNIT_ASSERT_VALUES_EQUAL(
+            KqpSchemeExec(runtime, R"(
+                CREATE TABLE `/Root/table` (key int, value int, PRIMARY KEY (key))
+                WITH (PARTITION_AT_KEYS = (10, 20, 30));
+            )"),
+            "SUCCESS");
+
+        const auto tableId = ResolveTableId(server, sender, "/Root/table");
+        const auto shards = GetTableShards(server, sender, "/Root/table");
+        UNIT_ASSERT_VALUES_EQUAL(shards.size(), 4u);
+
+        RunUpsertWithArbiter(
+            runtime, tableId, shards,
+            /* txId */ 1000001,
+            // arbiter will send 6 readsets (3 decisions + 3 expectations)
+            // shards will send 2 readsets each (decision + expectation)
+            /* expectedReadSets */ 6 + 3 * 2);
+
+        Cerr << "========= Checking table =========" << Endl;
+        UNIT_ASSERT_VALUES_EQUAL(
```

Comment:
Everywhere in this file another construction is used:
```
            auto tableState = TReadTableState(server, MakeReadTableSettings("/Root/table-1")).All();
            UNIT_ASSERT_VALUES_EQUAL(tableState, expectedTableState);
```

I suggest using a single approach.

Reply:
ReadTable uses persistent snapshots and very heavyweight scans with extra waits for history to be committed. It should never be used unless that's what you're actually testing.

Reply:
Is it preferable to change all the previous cases to `KqpSimpleExec` ?

Reply:
Either that, or make helper functions that would read using EvRead.